src/Storages/UVLoop.h

```diff
@@ -30,7 +31,34 @@ class UVLoop : public boost::noncopyable
     ~UVLoop()
     {
         if (loop_ptr)
-            uv_loop_close(loop_ptr.get());
+        {
+            auto res = uv_loop_close(loop_ptr.get());
+
+            if (res == UV_EBUSY)
+            {
+                LOG_DEBUG(log, "Closing pending handles");
+                uv_walk(loop_ptr.get(), onUVWalkClosingCallback, nullptr);
+
+                /// Run the loop until there are no pending callbacks.
+                while (true)
+                {
+                    res = uv_run(loop_ptr.get(), UV_RUN_ONCE);
+                    if (res)
+                        LOG_DEBUG(log, "Waiting for pending callbacks to finish ({})", res);
+                    else
+                        break;
+                }
```

Comment:
can we get "stuck" here and prevent server from shutting down?

Reply:
Actually I do not think so, because I also added closing of the connection object before this loop is destructed, so the number of events must be fixed and new events should not arise afterwards.

Reply:
And also handler is destructed before the loop too (see comment [d3e26c2](https://github.com/ClickHouse/ClickHouse/pull/44897/commits/d3e26c28cb15a24b740b733514d8e2973641805c)), so noone else can use this loop.